### PR TITLE
[#6872] Timeline V2 (beta) initial commit 

### PR DESCRIPTION
**Pull request by ga-ram**
Tested for both synchronous and asynchronous calls.

### CODE REVIEW

**Comment by ga-ram** [comment]
Screen Shot 2020-06-04 at 12 58 32 PM

File: web/src/main/angular/src/app/core/components/transaction-list-bottom-contents/index.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ import { SyntaxHighlightPopupModule } from 'app/core/components/syntax-highlight
 import { TransactionListBottomContentsContainerComponent } from './transaction-list-bottom-contents-container.component';
 import { HelpViewerPopupModule } from 'app/core/components/help-viewer-popup';
 import { ServerErrorPopupModule } from 'app/core/components/server-error-popup';
+import {TransactionTimelineModuleV2} from "../transaction-timeline-v2";
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-list-bottom-contents/index.ts]: Use absolute path and single quotes for the consistency.

```suggestion
import { TransactionTimelineModuleV2 } from 'app/core/components/transaction-timeline-v2';
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-container-v2.component.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                    if (call[this.keyIndex.applicationName] != asyncRootInfo[1][asyncInfoLastIndex]) {
+                        isAsync = false;
+                    } else if (call[this.keyIndex.tab] <= asyncRootInfo[0][asyncInfoLastIndex]) {
+                        asyncRootInfo[0].pop();
+                        asyncRootInfo[1].pop();
+                        isAsync = false;
+                    }
+                }
+
+                if(call[this.keyIndex.apiType] === "ASYNC") {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-container-v2.component.ts]: Space after `if` statement
```suggestion
                if (call[this.keyIndex.apiType] === "ASYNC") {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-container-v2.component.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                    isAsync = true;
+                    asyncRootInfo[0].push(call[this.keyIndex.tab]);
+                    asyncRootInfo[1].push(call[this.keyIndex.applicationName]);
+                }
+
+                let count = isAsync? async[depth].length : data[depth].length;
+                if (count != 0) {
+                    let prevSibling = isAsync? async[depth][count-1] : data[depth][count-1];
+                    if(call[this.keyIndex.begin] === prevSibling[this.keyIndex.begin]) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-container-v2.component.ts]: Space after `if` statement
```suggestion
                    if (call[this.keyIndex.begin] === prevSibling[this.keyIndex.begin]) {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-container-v2.component.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                        if(isAsync) {
+                            async[depth].pop();
+                        } else {
+                            data[depth].pop();
+                        }
+                    }
+                }
+
+                if(isAsync) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-container-v2.component.ts]: Space after if statement
```suggestion
                if (isAsync) {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
+import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
+
+@Component({
+    selector: 'pp-transaction-timeline-v2',
+    templateUrl: './transaction-timeline-v2.component.html',
+    styleUrls: ['./transaction-timeline-v2.component.css']
+})
+
+export class TransactionTimelineComponentV2 implements OnInit {
+    @Input() keyIndex: { [key: string]: number};
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Remove space 
```suggestion
    @Input() keyIndex: {[key: string]: number};
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    @Input() startTime: number;
+    @Input() endTime: number;
+    @Input() barRatio: number;
+    @Output() outSelectTransaction: EventEmitter<string> = new EventEmitter();
+    @Input() rowData: any[];
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Change order and syntax
```suggestion
    @Input() rowData: any[];
    @Output() outSelectTransaction = new EventEmitter<string>();
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    colorSet: { [key: string]: string } = {};
+    constructor() {}
+    ngOnInit() {
+        this.rowData = this.rowData.filter(row => row.length > 0);
+    }
+    private calcColor(str: string): string {
+        if (!(str in this.colorSet)) {
+            const color = [];
+            let hash = 0;
+            for ( let i = 0 ; i < str.length ; i++ ) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Remove space
```suggestion
            for (let i = 0; i < str.length; i++) {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                hash = str.charCodeAt(i) + ((hash << 5) - hash);
+            }
+            for ( let i = 0 ; i < 3 ; i++ ) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Remove space
```suggestion
            for (let i = 0; i < 3; i++) {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        };
+    }
+    getStartTime(call: any): number {
+        return call[this.keyIndex.begin] - this.startTime;
+    }
+
+    getMarginLeft(call: any, i: number, j: number): number {
+        let startTime: number;
+        let offset = 0;
+        if((j === 0)) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Give space after `if` statement
```suggestion
        if ((j === 0)) {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        }
+        return ((Number(call[this.keyIndex.begin]) - startTime) * this.barRatio) + offset;
+    }
+
+    onSelectCall(call: any): void {
+        this.outSelectTransaction.emit(call[this.keyIndex.id]);
+    }
+    findParent(call:any, depth: number): number {
+        let ret = -1;
+        if(depth > 0) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Give space after `if` statement
```suggestion
        if (depth > 0) {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+            if(this.rowData[depth-1].length > 0) {
+                this.rowData[depth-1].forEach((candidate: any, index: number) => {
+                    if(candidate[this.keyIndex.id] === call[this.keyIndex.parentId]) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Give space after `if` statement
```suggestion
            if (this.rowData[depth-1].length > 0) {
                this.rowData[depth-1].forEach((candidate: any, index: number) => {
                    if (candidate[this.keyIndex.id] === call[this.keyIndex.parentId]) {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+                    }
+                });
+            }
+        }
+        return ret;
+    }
+
+    showApplicationName(call: any, depth: number): boolean {
+        let parentIndex = this.findParent(call, depth);
+        if((depth === 0) || (parentIndex === -1)) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Give space after `if` statement
```suggestion
        if ((depth === 0) || (parentIndex === -1)) {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        if (this.rowData[depth - 1][parentIndex][this.keyIndex.applicationName] === call[this.keyIndex.applicationName]) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+    showName(call: any): boolean {
+        if(this.getWidth(call) > 180) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Give space after `if` statement
```suggestion
        if (this.getWidth(call) > 180) {
```

File: web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+            return true;
+        } else {
+            return false;
+        }
+    }
+    getText(call: any, depth: number): string {
+        let ret: string = "";
+        if(this.showApplicationName(call, depth)) {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-timeline-v2/transaction-timeline-v2.component.ts]: Give space after `if` statement
```suggestion
        if (this.showApplicationName(call, depth)) {
```

File: web/src/main/java/com/navercorp/pinpoint/web/view/TransactionInfoViewModel.java [hunk 15]
```diff
@@ -163,6 +163,10 @@ public String getDisableButtonMessage() {
                 }
                 first = false;
             }
+
+            if((record.getApiType() != "ASYNC") && (record.getElapsed() == 0)) {
```
Comment (emeroad) [inline web/src/main/java/com/navercorp/pinpoint/web/view/TransactionInfoViewModel.java]:
```suggestion
            if (!"ASYNC".equals(record.getApiType()) && (record.getElapsed() == 0)) {
```

Use the string.equals(),  not the == operator

**Comment by ga-ram** [comment]
async calls are shown as striped rectangles

Screen Shot 2020-06-16 at 5 39 52 PM

**Pull request merged**